lib/lhm/sql_retry.rb

```diff
@@ -15,22 +15,124 @@ module Lhm
+      old_prefix =  @log_prefix
+      @log_prefix = override_prefix unless override_prefix.nil?
+
       Retriable.retriable(retry_config) do
-        yield(@connection)
+        if @with_consistent_host
+          raise Lhm::Error.new("Could not reconnected to initial MySQL host. Aborting to avoid data-loss") unless same_host_as_initial?
+        end
+        # The time check will log a warning if an action takes more than 5s (there's a possibility a
+        # failover might've happened after the hostname check above)
+        with_time_check do
+          yield(@connection)
+        end
+      rescue Lhm::Error
+        # Since Lhm::Error < StandardError, it is important to separate the flows, otherwise every
+        # error would trigger a reconnect. In this case, some instances of Lhm::Error are raised in a controlled manner
+        # to trigger retries from ` Retriable.retriable`
+        raise
+      rescue StandardError => e
+        raise e unless error_can_trigger_reconnect?(e)
+        reconnect_with_host_check! if @with_consistent_host
       end
+    ensure
+      # Restore the initial log prefix once outside of the block (ex: Chunker -> ChunkInsert).
+      @log_prefix = old_prefix
     end
 
     attr_reader :retry_config
 
     private
 
+    def with_time_check
+      t_start = Time.new
+
+      # Captures return value from the provided block
+      return_value = yield
+
+      if ( Time.new - t_start) >= ABNORMAL_EXECUTION_TIME_THRESHOLD
+      log_with_prefix("Query took abnormal amount of time to execute and the host check might not be accurate anymore", :warn)
+      end
+
+      return_value
+    end
+
+    def hostname
+      # Context Should be defined by library caller
+      @connection&.execute("SELECT @@global.hostname").to_a.first.tap do |record|
+        return record&.first
+      end
+    end
+
+    def server_id
+      # This should only be used with CloudSQLs as their hosts are "localhost", but the "server_id" is different
+      @connection&.execute("SELECT @@global.server_id").to_a.first.tap do |record|
+        return record&.first
+      end
+    end
+
+    def log_with_prefix(message, level = :info)
+      message.prepend("[#{@log_prefix}] ") if @log_prefix
+      Lhm.logger.send(level, message)
+    end
+
+    def reconnect_with_host_check!
+      log_with_prefix("Lost connection to MySQL, will retry to connect to same host")
+      begin
+        Retriable.retriable(@host_retry_config) do
+          # tries to reconnect. On failure will trigger a retry
+          @connection.reconnect!
+          new_host = hostname
+          if new_host == @initial_host
+            # This is not an actual error, but it needs to trigger the Retriable
+            # from #with_retries to execute the desired logic again
+            raise Lhm::Error.new("LHM successfully reconnected to initial host: #{@initial_host}")
```

Comment:
If this is not an error, can we just not raise an error? Instead we can raise an internal error..?